thinc/backends/ops.py

```diff
@@ -680,6 +682,167 @@ def backprop_relu(
+        if inplace:
+            X *= 0.2  # type: ignore
+            X += 0.5  # type: ignore
+            return cast(FloatsType, self.xp.clip(X, 0, 1, out=X))
+        out = X * 0.2 + 0.5
+        return cast(FloatsType, self.xp.clip(out, 0, 1))
+
+    def backprop_hard_sigmoid(
+        self, dY: FloatsType, X: FloatsType, inplace: bool = False
+    ) -> FloatsType:
+        slope = self.alloc1f(1) + 0.2
+        dX = slope * ((-2.5 < X) & (X < 2.5))
+        if inplace:
+            dY *= dX
+            return dY
+        return dY * dX
+
+    def swish(self, X: FloatsType, inplace: bool = False) -> FloatsType:
+        if inplace:
+            X *= self.sigmoid(X)  # type: ignore
+            return cast(FloatsType, X)
+        out = X * self.sigmoid(X)  # type: ignore
+        return cast(FloatsType, out)
+
+    def backprop_swish(
+        self, dY: FloatsType, X: FloatsType, Y: FloatsType, inplace: bool = False
+    ) -> FloatsType:
+        Y = Y + self.sigmoid(X) * (1 - Y)  # type: ignore
+        if inplace:
+            dY *= Y  # type: ignore
+            return cast(FloatsType, dY)
+        out = dY * Y  # type: ignore
+        return cast(FloatsType, out)
+
+    # Following https://www.scitepress.org/Papers/2019/74696/74696.pdf
+    def hard_swish(self, X: FloatsType, inplace: bool = False) -> FloatsType:
+        if inplace:
+            X *= self.hard_sigmoid(X)  # type: ignore
+            return cast(FloatsType, X)
+        out = X * self.hard_sigmoid(X)  # type: ignore
+        return cast(FloatsType, out)
+
+    def backprop_hard_swish(
+        self, dY: FloatsType, X: FloatsType, inplace: bool = False
+    ) -> FloatsType:
+        dX = X * 0.4 + 0.5
+        dX[X > 2.5] = 1.0
+        dX[X < -2.5] = 0
+        if inplace:
+            dY *= dX
+            return dY
+        return dY * dX
+
+    # From https://arxiv.org/pdf/1905.02244v5.pdf
+    def hard_swish_mobilenet(self, X: FloatsType, inplace: bool = False) -> FloatsType:
+        if inplace:
+            X *= self.relu_n(X + 3) / 6
+            return X
+        return X * (self.relu_n(X + 3) / 6)
+
+    def backprop_hard_swish_mobilenet(
+        self, dY: FloatsType, X: FloatsType, inplace: bool = False
+    ) -> FloatsType:
+        dX = (1 / 6) * (X * 2.0 + 3.0)
+        dX[X > 3.0] = 1.0
+        dX[X < -3.0] = 0
+        if inplace:
+            dY *= dX
+            return dY
+        return dX * dY
+
+    # https://newbedev.com/is-there-an-easily-available-implementation-of-erf-for-python
+    def erf(self, X: FloatsType) -> FloatsType:
+        # save the sign of x
+        sign = self.xp.sign(X)
+        X = self.xp.abs(X)
+
+        a1 = 0.254829592
+        a2 = -0.284496736
+        a3 = 1.421413741
+        a4 = -1.453152027
+        a5 = 1.061405429
+        p = 0.3275911
+
+        t = 1.0 / (1.0 + p * X)
+        y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * self.xp.exp(
+            -X * X
+        )
+        return sign * y
```

Comment:
I think we'd need copyright attribution for such a large fragment?

If we want a naive implementation that we override with `NumpyOps`, I would use something like (untested):

```
erf = np.vectorize(lambda x: math.erf(x))
return erf(X)
```

Reply:
```
erf = np.vectorize(math.erf)
```

might also work

Reply:
Does cupy support `vectorize`? If not we can't do it that way in the `ops` function so easily.

Reply:
cupy has `vectorize`, but it doesn't know how to deal with `math` functions.

But does every `Ops` method have to run with CuPy arrays? Overriding this in `CupyOps` should be straightforward, since CuPy does have an `erf` function.

Reply:
I thought that the general design is that `ops` has the basic implementations that run both with `cupy` and `numpy` and then the `NumpyOps` and `CupyOps` can optimize it further. Is it hard to credit the website for the`erf` implementation somehow?

Reply:
Resolved by giving credit to John D. Cook, who put this code in the public domain (see comments attached to the article).